// @framer/brunosantos.loupe-slider-afternoon/code/Slider.tsx
import * as React from "react"
import { Frame, MotionValue, useMotionValue } from "framer"
//@ts-ignore
import { colors } from "@framer/aroagb.loupe-colors/code/canvas"

export function Slider(props) {
    // function onDirectionLock(axis) {
    //     console.log(axis)
    // }
    const circleSize = 15

    const [value, setValue] = React.useState(0)

    const boundRef = React.useRef<HTMLDivElement>()
    const knobRef = React.useRef<HTMLDivElement>()

    const updateValue = () => {
        if (!knobRef.current) {
            setValue(0)
        }

        setValue(Math.round(knobRef.current.getBoundingClientRect().right))
    }

    return (
        <Frame
            center
            ref={boundRef}
            style={{
                background: colors.primaryActive,
                height: 3,
                radius: 3,
            }}
            width="100%"
        >
            {/* fill */}
            <Frame
                style={{
                    background: colors.primary,
                    radius: 3,
                }}
                height="100%"
                width={value}
            ></Frame>
            {/* knob */}
            <Frame
                ref={knobRef}
                size={circleSize}
                center="y"
                style={{
                    background: colors.primary,
                    borderRadius: "50%",
                    boxShadow: "0px 2px 5px 0px rgba(0, 0, 0, 0.25)",
                }}
                drag="x"
                dragElastic={0}
                dragMomentum={false}
                dragConstraints={boundRef}
                onDrag={updateValue}
            ></Frame>
        </Frame>
    )
}

// Change the width and height here to fit your component
Slider.defaultProps = {
    width: 260,
    height: 44,
}
