// @framer/sebcglbailey.progress-ring/code/Progress_Ring.tsx

import * as React from "react";
import { Frame, animate, PropertyControls, ControlType, FramerAnimation, PropertyStore, RenderTarget } from "framer"

interface Props {
  value: number
  color: string
  width: number
  height: number
  margin: number
  stroke: number
  round: boolean
}

interface State {
  value: number
  margin: number
  stroke: number
  color: string
  round: boolean
}

const svgStyle: React.CSSProperties = {
  position: "absolute",
  transform: "rotate(-90deg)",
  overflow: "visible"
}

export class ProgressRing extends React.Component<Partial<Props>, State> {
  static defaultProps = {
    width: 300,
    height: 300,
    color: "#ffffff",
    value: 50,
    margin: 8,
    stroke: 2,
    round: false
  }

  state = {
    value: 370,
    margin: 8,
    color: "#ffffff",
    stroke: 2,
    round: false
  }

  static propertyControls: PropertyControls<Props> = {
    round: { type: ControlType.Boolean, title: "Rounded" },
    stroke: { type: ControlType.Number, title: "Width" },
    value: { type: ControlType.Number, title: "Value" },
    margin: { type: ControlType.Number, title: "Margin" },
    color: { type: ControlType.Color, title: "Color" }
  }

  componentDidMount() {
    const { value, margin} = this.props
    this.setState({ value, margin })
  }

  componentWillReceiveProps(props: Props) {
    let {value, margin, color, stroke, round } = this.props
    if (props.value !== value) {
      value = props.value
    }
    if (props.margin !== margin) {
      margin = props.margin
    }
    if (props.color !== color) {
      color = props.color
    }
    if (props.stroke !== stroke) {
      stroke = props.stroke
    }
    if (props.round !== round) {
      round = props.round
    }
    this.setState({value, margin, color, stroke, round})
  }

  render() {
    const { width, height } = this.props
    const { value, margin, color, stroke, round } = this.state

    const circ = Math.PI * (width - margin*2)
    const dashArray = circ
    const dashOffset = circ - (value/100) * circ

    return (
      <Frame
        width={width}
        height={height}
        background={null}
      >
        <svg
          style={{ ...svgStyle }}
          width={width}
          height={height}
        >
          <circle
            cx={width/2}
            cy={height/2}
            r={width/2 - margin}
            stroke-width={stroke} 
            stroke={`${color}`}
            strokeDasharray={circ}
            strokeDashoffset={dashOffset}
            strokeLinecap={round ? "round" : null}
            fill={"none"}
          />
        </svg>
      </Frame>
    )
  }
}
