// @framer/lily.baseui/code/navigation/Tabs.tsx
import * as System from 'baseui/tabs';
import {addPropertyControls, ControlType, RenderTarget, Frame} from 'framer';
import * as React from 'react';
import {cloneElement, createElement, useMemo, useEffect, useCallback} from 'react';
import {controls, merge} from '../generated/Tabs';
import {useManagedState} from '../utils/useManagedState';
import {placeholderState} from '../utils/placeholderState';
import {withHOC} from '../withHOC';
import * as icons from '../icons/utils';

const overrides = {
  TabContent: {style: {padding: '0px'}},
};

const Icon = icons.Menu;

const InnerTabs: React.SFC<any> = ({
  content,
  tabStyle,
  tabIcons,
  tabIconSize,
  activeKey: defaultActiveKey,
  ...props
}) => {
  const [activeKey, setActiveKey] = useManagedState(String(defaultActiveKey));
  const tabs = useMemo(() => {
    return content.map((tab, i) => {
      let title = tab.props.name;
      if (tabStyle === 'icon' && tabIcons[i] && tabIcons[i] !== 'none') {
        const Icon = icons[tabIcons[i]];
        if (Icon) {
          title = <Icon size={tabIconSize} />;
        }
      }

      return (
        <System.Tab title={title} key={String(i)}>
          {cloneElement(tab, {
            ...tab.props,
            style: {
              position: 'relative',
              width: '100%',
            },
          })}
        </System.Tab>
      );
    });
  }, [content]);

  const onChange = useCallback(({activeKey}) => {
    setActiveKey(String(activeKey));
  }, []);

  useEffect(() => setActiveKey(String(defaultActiveKey)), [defaultActiveKey]);

  return content.length === 0 &&
    (RenderTarget.current() === RenderTarget.canvas || RenderTarget.current() === RenderTarget.thumbnail) ? (
    createElement(placeholderState, {
      title: 'No tabs',
      label: 'Add tabs by connecting them on the Canvas',
    })
  ) : (
    <Frame size="100%" background="transparent" style={{overflow: 'hidden'}}>
      <System.Tabs onChange={onChange} overrides={overrides} {...props} activeKey={activeKey}>
        {tabs}
      </System.Tabs>
    </Frame>
  );
};

export const Tabs = withHOC(InnerTabs);

Tabs.defaultProps = {
  width: 500,
  height: 200,
};

addPropertyControls(Tabs, {
  activeKey: merge(controls.activeKey, {
    defaultValue: 0,
    type: ControlType.Number,
    min: 0,
    step: 1,
    displayStepper: true,
  }),
  renderAll: merge(controls.renderAll, {}),
  disabled: merge(controls.disabled, {}),
  orientation: merge(controls.orientation, {}),
  content: {
    title: 'Tabs',
    type: ControlType.Array,
    propertyControl: {
      type: ControlType.ComponentInstance,
    },
    defaultValue: [],
  },
  tabStyle: {
    title: 'Tab Style',
    type: ControlType.SegmentedEnum,
    options: ['text', 'icon'],
    optionTitles: ['Text', 'Icon'],
    defaultValue: 'text',
  },
  tabIcons: {
    title: 'Tab Icons',
    type: ControlType.Array,
    propertyControl: {
      type: ControlType.Enum,
      options: ['none', ...Object.keys(icons)],
    },
    defaultValue: [],
    hidden: props => props.tabStyle !== 'icon',
  },
  tabIconSize: {
    type: ControlType.Number,
    title: 'Tab Icon Size',
    defaultValue: 20,
    hidden: props => props.tabStyle !== 'icon',
  },
});
